### ISSUE TITLE
Pre-commit.pyz Bump to Python3.12

### ISSUE THREAD

**Issue by iPoetDev** [pre-commit/pre-commit#3090]
### search you tried in the issue tracker

pyz, 3.12 . 312

### describe your issue

I am using `pre-commit.pyz` as a standalone. (I do not want a `venv` nor need one). I do want to run pre-commit as a local runner and pre-commit task/hook.

I migrated to Python 3.12 and uninstalled Python311. (No, I have not mastered dual use python versions, yet; I watch your YouTubes)
- I only have Python312 installed
- Is installed on system Path
- python launches the python repl for 3.12.

I follow this excellent article to run Pre-commit as standalone (intent: a simple nodejs/native-js project, no need for venv): https://pakstech.com/blog/pre-commit-anywhere/

`PS D:\Code\CodeInstitute\examples\AlpineTicTacToe> python C:\users\<$USER>\bin\pre-commit.pyz run --all-files`

```cmd
[WARNING] Unexpected key(s) present at root: pre-commit.ci
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/editorconfig-checker/editorconfig-checker.python.
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('C:\\Users\\<$USER>\\.cache\\pre-commit-zipapp\\hkKBCoxnABE0STlcNAKzB3rewvTxFyzlQoAGhB9XiP8\\python.exe', '-mvirtualenv', 'C:\\Users\\<$USER>\\.cache\\pre-commit\\repodbi8x3dg\\py_env-python3.11', '-p', 'python3.11')
return code: 1
stdout:
    RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.11'
stderr: (none)
Check the log at C:\Users\Charles\.cache\pre-commit\pre-commit.log
```
Log is attached

--- 
So I scratched around in the 
```
C:\Users\<$USER>\.cache\pre-commit\repodbi8x3dg
```
and I found that the setup.cfg has

```
[options]
packages = find:
install_requires =
    ruamel.yaml>=0.15
    tomli>=1.1.0;python_version<"3.11"
python_requires = >=3.8
```

Question: If I change `python_version<"3.11"` to `python_version<"3.12"` will this work (untested)?

[pre-commit.log](https://github.com/pre-commit/pre-commit/files/13625071/pre-commit.log)

### pre-commit --version

@latest

### .pre-commit-config.yaml

```yaml
pre-commit.ci:
  autofix_commit_msg: |
    chore [pre-commit.ci] commit: auto fixes from pre-commit.ci hooks
  autoupdate_commit_msg: |
    chore [pre-commit.ci] pre-commit autoupdate
  autofix_prs: true
  autoupdate_branch: main
  autoupdate_schedule: monthly
  push:
    branches:
      - main
  skip:
    - editorconfig-checker
    - trailing-whitespace
    - end-of-file-fixer
    - mixed-line-ending
    - check-yaml
  submodules: false
fail_fast: false
default_language_version:
  python: python3.11
repos:
  - repo: local
    hooks:
      - id: add-pre-commit-config
        name: Add pre-commit config
        entry: git add .pre-commit-config.yaml
        language: system
        pass_filenames: false
        stages: [commit]
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.5.0
    hooks:
      - id: trailing-whitespace
        # exclude: ^staticfiles/admin/|^docs/
      - id: check-yaml
        exclude: ^pep8speaks\.yml$
      - id: end-of-file-fixer
        # exclude: ^staticfiles/|^docs/
      - id: double-quote-string-fixer
      - id: detect-private-key
      - id: mixed-line-ending
      - id: no-commit-to-branch
        args: [--branch, main]
  - repo: https://github.com/editorconfig-checker/editorconfig-checker.python
    rev: 2.7.3
    hooks:
      - id: editorconfig-checker
        alias: ec
#  - repo: https://github.com/Yelp/detect-secrets
#    rev: v1.0.3
#    hooks:
#      - id: detect-secrets
#        args: ['--baseline', 'secrets.baseline']
#        exclude: 'secrets.baseline'
```

### version information

```
pre-commit version: 3.5.0
git --version: git version 2.43.0.windows.1
sys.version:
    3.12.0 (tags/v3.12.0:0fb18b0, Oct  2 2023, 13:03:39) [MSC v.1935 64 bit (AMD64)]
sys.executable: C:\Users\<$USER>\.cache\pre-commit-zipapp\hkKBCoxnABE0STlcNAKzB3rewvTxFyzlQoAGhB9XiP8\python.exe
os.name: nt
sys.platform: win32
```

### error information

```
An unexpected error has occurred: CalledProcessError: command: ('C:\\Users\\<$USER>\\.cache\\pre-commit-zipapp\\hkKBCoxnABE0STlcNAKzB3rewvTxFyzlQoAGhB9XiP8\\python.exe', '-mvirtualenv', 'C:\\Users\\Charles\\.cache\\pre-commit\\repodbi8x3dg\\py_env-python3.11', '-p', 'python3.11')
return code: 1
stdout:
    RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.11'
stderr: (none)
```

```
Traceback (most recent call last):
  File "C:\Users\<$USER>\.cache\pre-commit-zipapp\hkKBCoxnABE0STlcNAKzB3rewvTxFyzlQoAGhB9XiP8\wheels\pre_commit-3.5.0-py2.py3-none-any.whl\pre_commit\error_handler.py", line 73, in error_handler
    yield
  File "C:\Users\<$USER>\.cache\pre-commit-zipapp\hkKBCoxnABE0STlcNAKzB3rewvTxFyzlQoAGhB9XiP8\wheels\pre_commit-3.5.0-py2.py3-none-any.whl\pre_commit\main.py", line 414, in main
    return run(args.config, store, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Charles\.cache\pre-commit-zipapp\hkKBCoxnABE0STlcNAKzB3rewvTxFyzlQoAGhB9XiP8\wheels\pre_commit-3.5.0-py2.py3-none-any.whl\pre_commit\commands\run.py", line 442, in run
    install_hook_envs(to_install, store)
  File "C:\Users\<$USER>\.cache\pre-commit-zipapp\hkKBCoxnABE0STlcNAKzB3rewvTxFyzlQoAGhB9XiP8\wheels\pre_commit-3.5.0-py2.py3-none-any.whl\pre_commit\repository.py", line 248, in install_hook_envs
    _hook_install(hook)
  File "C:\Users\<$USER>\.cache\pre-commit-zipapp\hkKBCoxnABE0STlcNAKzB3rewvTxFyzlQoAGhB9XiP8\wheels\pre_commit-3.5.0-py2.py3-none-any.whl\pre_commit\repository.py", line 95, in _hook_install
    lang.install_environment(
  File "C:\Users\<$USER>\.cache\pre-commit-zipapp\hkKBCoxnABE0STlcNAKzB3rewvTxFyzlQoAGhB9XiP8\wheels\pre_commit-3.5.0-py2.py3-none-any.whl\pre_commit\languages\python.py", line 212, in install_environment
    cmd_output_b(*venv_cmd, cwd='/')
  File "C:\Users\<$USER>\.cache\pre-commit-zipapp\hkKBCoxnABE0STlcNAKzB3rewvTxFyzlQoAGhB9XiP8\wheels\pre_commit-3.5.0-py2.py3-none-any.whl\pre_commit\util.py", line 110, in cmd_output_b
    raise CalledProcessError(returncode, cmd, stdout_b, stderr_b)
pre_commit.util.CalledProcessError: command: ('C:\\Users\\<$USER>\\.cache\\pre-commit-zipapp\\hkKBCoxnABE0STlcNAKzB3rewvTxFyzlQoAGhB9XiP8\\python.exe', '-mvirtualenv', 'C:\\Users\\Charles\\.cache\\pre-commit\\repodbi8x3dg\\py_env-python3.11', '-p', 'python3.11')
return code: 1
stdout:
    RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.11'
stderr: (none)
```

**Comment by iPoetDev** [comment]
I spotted that my pre-commit-config.yaml has 3.11 for python too...

**Comment by iPoetDev** [comment]
I reran the PS D:\Code\CodeInstitute\examples\AlpineTicTacToe> python C:\users\<$USER>\bin\pre-commit.pyz run --all-files after changing the setup.cfg and it was successful.

Minor concern, edge use case, still worthy.

**Comment by asottile** [comment]
this isn't a pre-commit problem -- the `.pyz` is version indepdendent -- you've asked for `python3.11` in your configuration

**Comment by iPoetDev** [comment]
No, the pyz cfg is pins to max of 3.11

Me setting it to 3.12 worked.

**Comment by asottile** [comment]
the pyz has no configuration in it

**Comment by iPoetDev** [comment]
![2023-12-09 22_01_02-repodbi8x3dg](https://github.com/pre-commit/pre-commit/assets/51715025/83266310-9928-4b4e-bf89-fc5f53c8fe92)

**Comment by iPoetDev** [comment]
I am just providing my experience and what worked.

**Comment by asottile** [comment]
that has nothing to do with your problem and is not part of the pyz.

your problem is that you have `default_language_version: python3.11` in **your configuration file**

**you** explicitly asked pre-commit to use python 3.11

**Comment by iPoetDev** [comment]
Fine. It worked; regardless.

**Comment by asottile** [comment]
your edit that you made in `setup.cfg` **did not fix the problem**.  that file is python packaging metadata for pre-commit/pre-commit-hooks that instructs pip on what dependencies it needs to install

**Comment by iPoetDev** [comment]
Thank you. As a user, I am not the author., so I have to use the clues in in front of me (in error codes) a as first contact pain.

:\\Users\\<$USER>\\.cache\\pre-commit\\`repodbi8x3dg`\\py_env-python3.11' so I looked into the folder in the error and changed the cfg 

and then logging this ticket, I spotted the config.

still .. I removed all references of 3.11 from computer and it was either a config of setup or my yaml. 

It works now.
